### PR TITLE
JDK-8255978: [windows] os::release_memory may not release the full range

### PR DESCRIPTION
**Pull request by tstuefe**
On Windows, os::release_memory(p,size) may not actually release the whole region if it contains multiple mappings. This may cause memory bloat or runaway leaks or errors which look like failed mappings to specific wish addresses.

Background:

On Windows, memory mappings are established with VirtualAlloc() [1] and released with VirtualFree() [2]. In constrast to POSIX munmap(), VirtualFree() can only release a single full range established with VirtualAlloc(). It cannot release multiple ranges, or parts of a range.

The Windows implementation of os::release_memory(p, size) [3] calls VirtualFree(p, NULL, MEM_RELEASE) - it ignores the size parameter and releases whatever mapping happens to start at p:

```
bool os::pd_release_memory(char* addr, size_t bytes) {
  return VirtualFree(addr, 0, MEM_RELEASE) != 0;
}
```

... which assumes that the given range size corresponds to the size of a mapping starting at p.

This may be incorrect:

1) For NUMA-friendly allocation, we allocate memory in stripes, each stripe individually allocated.
2) For +UseLargePagesIndividualAllocation we do the same
3) apart from that, the given region size may just be wrong. Since we never check these, we may never have noticed. I am currently running tests to find out if we have other mismatched releases.

For cases (1) and (2), we would just release the first stripe in that striped range, leaving the rest of the mappings intact. This is not immediately noticeable, since VirtualFree() returns success. And even if it did not, we usually ignore the return code of os::release_memory().

The problem is aggrevated since, on Windows, we often employ an "optimistically-release-and-remap" approach: since mappings are undivisible, if one wants to change their size, split them or similar, one has to follow this sequence:

a) release old allocation
b) place into the now vacated address room one or more new allocations

This is not guaranteed to work, since between (a) and (b) someone may have grabbed the address space. We live with that since there is no way to do this differently.

When used on a range which contains multiple mappings, this technique is almost guaranteed to fail. In that case, (a) would only release the first mapping in the range. (b) would almost certainly fail since most of the original range would still be mapped.

Examples of these technique in os_windows.cpp:
- os::split_reseved_memory() (see also [4])
- map_or_reserve_memory_aligned()
- os::replace_existing_mapping_with_file_mapping()

This can manifest as small memory leak or inability to attach to a given wish address. It could also result in a viscous loop ([5], [6]) and result in ballooning and native OOMs.

--

AFAICS this is an old issue, dating back to at least jdk 8.

--

The change in detail:

- os::release_memory() on Window now:
	- in LP/NUMA case we unmap all mappings in the given range. The mappings must match the range exactly, otherwise we assert.
	- In normal case we just check the given range and assert if it does not match.
	Note: in release builds I just print a warning and return false. Should this be a guarantee? I leave that up to the Reviewers.

- Added gtests which test os::release_memory() with a variety of scenarios. Includes both positive and negative tests. The latter includes a death test for debug.

- Added a helper function to find mappings, for now windows only (os::win32::find_mapping()), including gtests for that function.

- Added a function to print mappings, for diagnostic purposes: os::print_memory_mappings(). Provided an implementation for linux and windows.

--
[1] https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc
[2] https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualfree
[3] https://github.com/openjdk/jdk/blob/5dfb42fc68099278cbc98e25fb8a91fd957c12e2/src/hotspot/os/windows/os_windows.cpp#L3394
[4] https://bugs.openjdk.java.net/browse/JDK-8253649
[5] https://github.com/openjdk/jdk/blob/5dfb42fc68099278cbc98e25fb8a91fd957c12e2/src/hotspot/os/windows/os_windows.cpp#L3150
[6] https://bugs.openjdk.java.net/browse/JDK-8255954

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ❌ (1/9 failed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

**Failed test task**
- [Linux x64 (hs/tier1 runtime)](https://github.com/tstuefe/jdk/runs/1423207838)

### Issue
 * [JDK-8255978](https://bugs.openjdk.java.net/browse/JDK-8255978): [windows] os::release_memory may not release the full range

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 0a47110ba9b558c5fcd99e7a6040940bb66f5668
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**) ⚠️ Review applies to c999120035527e87b114bc324141131693aac471

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1143/head:pull/1143`
`$ git checkout pull/1143`

### CODE REVIEW

**Review by iklam — APPROVED** [review]
LGTM. I am running on mach5 tiers 1-3 to make sure the new test cases don't have any surprises :-)

**Comment by tstuefe** [comment]
> LGTM. I am running on mach5 tiers 1-3 to make sure the new test cases don't have any surprises :-)

Thank you Ioi!

Our nightlies yesterday did not show anything, but we do not test much with LP or NUMA. Nor with NVDimm. So exotic combinations like those may pass us by.

**Comment by iklam** [comment]
> LGTM. I am running on mach5 tiers 1-3 to make sure the new test cases don't have any surprises :-)

mach5 tiers 1-3 finished with no failure.

**Comment by tstuefe** [comment]
Great, thanks Ioi.

May I please have a second review? This is tied to a P2 issue.

**Comment by tstuefe** [comment]
Gentle ping.

File: test/hotspot/gtest/runtime/test_os.cpp
Comment (yminqi) [inline test/hotspot/gtest/runtime/test_os.cpp:353]: Can this comment be removed?

File: test/hotspot/gtest/runtime/test_os.cpp
Comment (yminqi) [inline test/hotspot/gtest/runtime/test_os.cpp:440]: The comments should be aligned.

**Review by yminqi — APPROVED** [review]
LGTM, very nice test case, thanks!

**Comment by tstuefe** [comment]
> LGTM, very nice test case, thanks!

Thanks a lot Yumin!

I'll change the comment before pushing. I'd like to leave that printing stuff in, but will disable it by default. That way we can switch it on when needed to look what is going on.

**Comment by tstuefe** [comment]
I'll integrate now. The remaining error, a crash on Linux in HeapRegion::block_size,  has nothing to do with this patch (I opened https://bugs.openjdk.java.net/browse/JDK-8256641 to track it).

**Pull request closed without merging**
